Memcached as a Service Support

oza opened the pull request:
Memcached is a de fact standard in-memory caching system.
This patch provides Cloud Foudnry with the provisioning, unprovisioning Memcached instances.

And, happy new year!

Signed-off-by: OZAWA Tsuyoshi ozawa.tsuyoshi@gmail.comDear Cloud Foundry contributor,

<p>
If you are reading this message, it means you submitted a pull request in the
Cloud Foundry GitHub repository.
</p>


<p>
First of all, thanks! We really appreciate your participation.
</p>


<p>
Recently we made some changes in how we are verifying and reviewing open source
contributions like yours. In addition, we changed the way we can expose our
internal development in real-time. The changes are exciting, as they allow all
our staff to collaborate seamlessly with you, which increases our mutual
velocity and gives the community a bigger stake in our direction.
</p>


<p>
The Cloud Foundry team uses Gerrit, a code review tool that originated in the
Android Open Source Project. We also use GitHub as an official mirror, though
all pull requests are accepted via Gerrit.
</p>


<p>
Follow these steps to make a contribution to any of our open source
repositories:
</p>

1. Complete our CLA Agreement for
   [individuals](http://www.cloudfoundry.org/individualcontribution.pdf) or
   [corporations](http://www.cloudfoundry.org/corpcontribution.pdf).
2. Sign up for an account on our public Gerrit server at
   http://reviews.cloudfoundry.org/.
3. Create and upload your public SSH key in your Gerrit account profile.
4. Set your name and email:
   
   ```
           git config --global user.name "Firstname Lastname"
           git config --global user.email "your_email@youremail.com"
   ```
5. Install our gerrit-cli gem:
   
   ```
           gem install gerrit-cli
   ```
6. Clone the Cloud Foundry repo:
   _Note: to clone the BOSH repo, or the Documentation repo, replace
   `vcap` with `bosh` or `oss-docs`_
   
   ```
           gerrit clone ssh://reviews.cloudfoundry.org:29418/vcap
           cd vcap
   ```
7. Make your changes, commit, and push to gerrit:
   
   ```
           git commit
           gerrit push
   ```

<p>
Once your commits are approved by our Continuous Integration Bot (CI Bot) as
well as our engineering staff, return to the Gerrit interface and MERGE your
changes. The merge will be replicated to GitHub automatically at
https://github.com/cloudfoundry. If you get feedback on your submission, we
recommend squashing your commit with the original change-id. See the squashing
section here for more details: https://help.github.com/rebase.
</p>

igrigorik commented on memcached/lib/memcached_service/memcached_error.rb:
```diff
@@ -0,0 +1,27 @@
+        # 31100 - 31199  Memcached-specific Error
+        # FIXME: This error code is paste from redis.
+        MEMCACHED_SAVE_INSTANCE_FAILED        = [31100, HTTP_INTERNAL, "Could not save instance: %s"]
+        MEMCACHED_DESTORY_INSTANCE_FAILED     = [31101, HTTP_INTERNAL, "Could not destroy instance: %s"]
```
Typo.. "DESTORY"

oza replied:
Thank you for your pointing out. I'll fix it.

krobertson commented:
+1. I was pondering writing one as well.

It looks like there is still some commented out parts referring to redis, and I do agree some authentication would be good with like SASL.

oza commented:
@yssk22r @krobertson Thank you for your advice. I'm going to support authentication with SASL and delete needless source code.

Commit:
Memcached as a Service Support

Memcached is a de fact standard in-memory caching system.
This patch provides Cloud Foundry with the provisioning,
unprovisioning Memcached instances.

Some advices are given by:

Ken Robertson <ken _at_ invalidlogic.com>
Yohei Sasaki  <yssk22 _at_ gmail.com>
Ilya Grigorik <ilya _at_ igvita.com>

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

Commit:
memcached_node: added SASL support.

In PaaS environment, the authentication feature is required.
This patch provides MaaS with SASL support of memcached.

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

natea commented:
Would be awesome to have memcached-as-a-service in CloudFoundry!

krobertson commented:
Looks good! Definitely going to pull this into my code tree for PaaS.io! Will report back once I test it out.

leto commented on memcached/lib/memcached_service/memcached_node.rb:
```diff
@@ -0,0 +1,617 @@
+    varz[:plan] = instance.plan
+    varz[:usage] = {}
+    varz[:usage][:max_memory] = instance.memory.to_f * 1024.0
+    # FIXME : get more detail information
```
Usage info will be necessary to merge this.

oza replied:
Fixed at https://github.com/nttcom/vcap-services/commit/1fff38ee132bf922726d76234f06952fa51b5661

leto commented on memcached/spec/node_spec.rb:
```diff
@@ -0,0 +1,486 @@
+      @instance.destroy
+    end
+
+    #it "should start a new instance if the instance is not started when start all provisioned instances" do
```
What is up with this?

oza replied:
Fixed at https://github.com/nttcom/vcap-services/commit/ee8325ee628354bddcb1e32bb9a20a5e20051fb2

krobertson commented on memcached/lib/memcached_service/memcached_node.rb:
```diff
@@ -0,0 +1,617 @@
+  def check_password(port, password)
+    Timeout::timeout(@memcached_timeout) do
+      #memcached = Dalli::Client.new({:port => port})
+      #memcached.auth(password)
```
Does this need to be fixed or uncommented?  Looks like it will always return true.

oza replied:
check_password() is the redis-specific function because of redis's not having user-level auth feature.
Memcached has it, so I deleted this code at : https://github.com/nttcom/vcap-services/commit/b07dbc8fb0d13a58a46963d10356d3b6097f1854

krobertson commented on memcached/lib/memcached_service/memcached_node.rb:
```diff
@@ -0,0 +1,617 @@
+      #memcached.auth(password)
+    end
+    true
+  rescue => e
```
If you want to handle timeouts properly, this should be `rescue Exception`.  A blank `rescue` only rescues a `StandardError` or below, but a `Timeout:Error` inherits from `Exception` instead.

http://rubylearning.com/satishtalim/ruby_exceptions.html

oza replied:
Thanks, fixed at : https://github.com/nttcom/vcap-services/commit/ee8325ee628354bddcb1e32bb9a20a5e20051fb2

krobertson commented on memcached/lib/memcached_service/memcached_node.rb:
```diff
@@ -0,0 +1,617 @@
+      raise MemcachedError.new(MemcachedError::MEMCACHED_NOT_YET_IMPLEMTED)
+    end
+    info
+  rescue => e
```
Ditto as above

oza replied:
This bug is fixed at  https://github.com/nttcom/vcap-services/commit/ee8325ee628354bddcb1e32bb9a20a5e20051fb2 , too :-)

krobertson commented on memcached/spec/node_spec.rb:
```diff
@@ -0,0 +1,486 @@
+
+    # FIXME: authentication feature
+    it "should not allow null credentials to access the instance" do
+      pending "authentication feature is not implemented."
```
Looks like a lot of the authentication based tests are still pending too

oza replied:
All test cases are passed at https://github.com/nttcom/vcap-services/commit/ee8325ee628354bddcb1e32bb9a20a5e20051fb2

oza commented:
@krobertson @leto Thank you for your review! I'll fix the parts you pointed out, and commit to the memcached branch.

Commit:
memcacehed_node: support to the stats of memcached instances

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

Commit:
memcacehed_node: supported get_healthz

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

Commit:
spec: added authentication tests.

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

Commit:
tests: Passed all test cases.

There are several fixes including:
	1. avoiding instances leak.
	2. authentication supports.
	3. SASLAdmin's handling empty inputs.

For more detail, Please see the source codes under the spec directory.

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

Commit:
memcached_node: deleted needless codes.

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

oza commented:
Now, all tasks are done. Please review the codes.

oza commented:
@leto Thanks :-)

oza commented:
@leto Please let me know if there are some points to fix to merge this patch.

krobertson commented:
Sorry, haven't had a chance yet to review it some more but definitely will be.

leto commented:
@oza we are transitioning to a new review system, if you are confident with the current patch, I can help push it through :) Have you ever used Gerrit?

oza commented:
@krobertson I'm looking forward to your reviewing :-)

oza commented:
@leto In fact, I've never used Gerrit. I'm going to learn it, however, if your new review system use it.

leto commented:
@oza yes, we will continue to mirror to Github, but very soon we will have a public Gerrit instance so that external and internal developers will be working together and using the same workflow.

oza commented:
@leto It's the great news for us! We're looking forward to joining to the exciting collaboration! Please let me know if it becomes available.

leto commented:
@oza My highest priority is getting a a public Gerrit instance working. I will definitely let you know when it is close to ready :)

Also, it is our policy to leave copyright lines alone and instead add people/companies who contribute to https://github.com/cloudfoundry/vcap/blob/master/AUTHORS . Could you undo the changes to the copyright lines in this PR and create a new PR for vcap that adds yourself/NTT to AUTHORS?

Once that happens I can push this into our internal Gerrit so we can enjoy memcached support before the public Gerrit is ready :)

Commit:
copyright: fixed copyrights.

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

Commit:
rspec: fixed typo and memcached instance leak

This commit includes 2 fixes:
1. Avoid memcached instance leak when running test cases.
2. Fixed some typo.

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

Commit:
memcached_node:start instance in enable_instance()

Fixed to call start_instance() in enable_instance().
This change allows MaaS to work rebalancing correctly.

Signed-off-by: UEMURA Yuichi <y.uemura@ntt.com>

krobertson commented on memcached/lib/memcached_service/memcached_node.rb:
```diff
@@ -0,0 +1,573 @@
+    true
+  end
+
+  def enable_instance(service_credentials, binding_credentials_map = {})
```
I noticed that in `disable_instance`, there is a call to `stop_instance`, however in `enable_instance`, there is no call to start it again.

oza replied:
@krobertson Thank you for your review!  This bug may cause unexpected case(especially when rebalancing), therefore I fixed at https://github.com/nttcom/vcap-services/commit/93a6a291881a4caacf5dda941838565286778e40

oza commented:
@leto OK.

At first, I fixed the copyrights for merging at https://github.com/nttcom/vcap-services/commit/eff7956c08fff9931fe70230ac77b9b36e3cd960 

Also, we'd like to add our name to the AUTHOR file, and should I send pull request to vcap itself instead of vcap-services?

leto commented:
@oza, yes since vcap is the main repo and all others are submodules inside of vcap, we just keep one AUTHORS file.

oza commented:
@leto We sent the pull request! Please check out: https://github.com/cloudfoundry/vcap/pull/185

oza commented:
@leto Please notify me if there appear tasks which I should do or I can help.

oza commented:
Wow, big changes came. We'll start to rebase to catch up vmware changes!
Please give me comments if there appear some fix points to merge.  > commiters
